File: backend-python/app/models/partner.py
"""Partner domain models — Sprint 5.2 (MongoDB integration).

These mirror `shared/src/types/partner.ts` byte-for-byte so the
partner-frontend consumes identical payloads from the mock TS server and
FastAPI + MongoDB.
"""

from __future__ import annotations

from typing import List, Literal, Optional

from pydantic import BaseModel

PartnerOrderStatus = Literal[
    "new", "accepted", "picked", "processing", "ready", "delivered", "cancelled"
]

BusinessCategory = Literal["laundry", "dry-clean", "premium", "shoe-care"]


class PartnerOrderItem(BaseModel):
    id: str
    name: str
    qty: int
    price: int


class PartnerOrderTimelineStage(BaseModel):
    id: str
    label: str
    time: str
    done: bool


class PartnerOrderResponse(BaseModel):
    id: str
    code: str
    customerName: str
    customerPhone: str
    status: PartnerOrderStatus
    placedAt: str
    slot: str
    address: str
    itemCount: int
    amount: int
    paymentMode: Literal["online", "cod"]
    serviceLabel: str
    items: List[PartnerOrderItem] = []
    timeline: List[PartnerOrderTimelineStage] = []


class RejectOrderPayload(BaseModel):
    reason: str = ""


class PartnerProfileResponse(BaseModel):
    partnerId: str
    businessName: str = "QuickPress Partner Store"
    ownerName: str = "Partner"
    phone: str = ""
    email: str = ""
    city: str = "Bengaluru"
    rating: float = 5.0
    totalOrders: int = 0
    joinedOn: str = "August 2026"
    onTimeRate: float = 98.5
    tier: Literal["Bronze", "Silver", "Platinum", "Gold"] = "Silver"
    isVerified: bool = False
    status: str = "pending"


class PartnerProfileUpdate(BaseModel):
    businessName: Optional[str] = None
    ownerName: Optional[str] = None
    phone: Optional[str] = None
    email: Optional[str] = None
    city: Optional[str] = None


class BusinessSettingsResponse(BaseModel):
    isStoreOpen: bool
    acceptingNewOrders: bool
    autoAcceptOrders: bool
    expressDelivery: bool
    pickupRadiusKm: int
    openingTime: str
    closingTime: str
    weeklyOff: str
    dailyOrderCap: int


class BusinessSettingsUpdate(BaseModel):
    isStoreOpen: Optional[bool] = None
    acceptingNewOrders: Optional[bool] = None
    autoAcceptOrders: Optional[bool] = None
    expressDelivery: Optional[bool] = None
    pickupRadiusKm: Optional[int] = None
    openingTime: Optional[str] = None
    closingTime: Optional[str] = None
    weeklyOff: Optional[str] = None
    dailyOrderCap: Optional[int] = None


class PartnerServiceResponse(BaseModel):
    id: str
    name: str
    unit: str
    price: int
    turnaroundHours: int
    enabled: bool
    category: BusinessCategory


class PartnerServiceCreate(BaseModel):
    name: str
    unit: str = "per item"
    price: int
    turnaroundHours: int = 24
    enabled: bool = True
    category: BusinessCategory = "laundry"


class PartnerServiceUpdate(BaseModel):
    name: Optional[str] = None
    unit: Optional[str] = None
    price: Optional[int] = None
    turnaroundHours: Optional[int] = None
    enabled: Optional[bool] = None
    category: Optional[BusinessCategory] = None


class PartnerDashboardResponse(BaseModel):
    newOrders: int
    inProgress: int
    readyForDelivery: int
    delivered: int
    earningsToday: int


class PartnerEarningsResponse(BaseModel):
    total: int
    orders: int


class PartnerWalletResponse(BaseModel):
    accountId: str
    balance: float
    cashbackBalance: float
    rewardPoints: int
    referralCode: str
    referralEarned: float
    onHold: float = 0
    lifetimeEarned: float = 0
    bankLast4: str = ""
    autoPayout: bool = True


class PartnerWalletTransactionResponse(BaseModel):
    id: str
    title: str
    date: str
    amount: float
    direction: Literal["credit", "debit"]
    status: Literal["success", "pending", "failed"]
    kind: str


class WithdrawPayload(BaseModel):
    amount: float


class PartnerReviewResponse(BaseModel):
    id: str
    partnerId: str
    customerName: str
    rating: float
    comment: str
    date: str


class PartnerNotificationResponse(BaseModel):
    id: str
    title: str
    body: str
    date: str
    read: bool
    kind: str


class OnboardingPayload(BaseModel):
    businessName: str
    ownerName: str
    category: BusinessCategory = "laundry"
    gstin: str = ""
    address: str = ""
    city: str = ""
    pincode: str = ""
    openingTime: str = "08:00"
    closingTime: str = "21:00"


class OnboardingResponse(BaseModel):
    partnerId: str
    phone: str
    businessName: str
    isVerified: bool
    isOnboarded: bool

File: admin-frontend/src/shared/types/partner.ts
// Partner app domain types. These mirror the shared QuickPress backend contracts
// consumed by both customer-frontend and partner-frontend.

export type PartnerAuthStep = "phone" | "otp" | "registration" | "ready";

export type PartnerSession = {
  partnerId: string;
  phone: string;
  businessName: string;
  isVerified: boolean;
  isOnboarded: boolean;
};

export type BusinessCategory = "laundry" | "dry-clean" | "premium" | "shoe-care";

export type BusinessRegistrationPayload = {
  businessName: string;
  ownerName: string;
  category: BusinessCategory;
  gstin: string;
  address: string;
  city: string;
  pincode: string;
  openingTime: string;
  closingTime: string;
};

export type PartnerOrderStatus =
  | "new"
  | "accepted"
  | "picked"
  | "processing"
  | "ready"
  | "delivered"
  | "cancelled";

export type PartnerOrderItem = {
  id: string;
  name: string;
  qty: number;
  price: number;
};

export type PartnerOrder = {
  id: string;
  code: string;
  customerName: string;
  customerPhone: string;
  status: PartnerOrderStatus;
  placedAt: string;
  slot: string;
  address: string;
  itemCount: number;
  amount: number;
  paymentMode: "online" | "cod";
  serviceLabel: string;
  items: PartnerOrderItem[];
  timeline: { id: string; label: string; time: string; done: boolean }[];
};

export type PartnerServiceRate = {
  id: string;
  name: string;
  unit: string;
  price: number;
  turnaroundHours: number;
  enabled: boolean;
  category: BusinessCategory;
};

export type EarningsPoint = {
  label: string;
  amount: number;
};

export type EarningsSummary = {
  today: number;
  week: number;
  month: number;
  pendingPayout: number;
  completedOrders: number;
  avgOrderValue: number;
  trend: EarningsPoint[];
  payouts: {
    id: string;
    date: string;
    amount: number;
    status: "paid" | "processing" | "failed";
    utr: string;
  }[];
};

export type PartnerWalletSummary = {
  availableBalance: number;
  onHold: number;
  lifetimeEarned: number;
  bankLast4: string;
  autoPayout: boolean;
};

export type PartnerWalletTransaction = {
  id: string;
  title: string;
  date: string;
  amount: number;
  direction: "credit" | "debit";
  status: "success" | "pending" | "failed";
  kind: "order-payout" | "settlement" | "penalty" | "incentive" | "withdrawal";
};

export type PartnerProfile = {
  partnerId: string;
  businessName: string;
  ownerName: string;
  phone: string;
  email: string;
  city: string;
  rating: number;
  totalOrders: number;
  joinedOn: string;
  onTimeRate: number;
  tier: "Bronze" | "Silver" | "Gold" | "Platinum";
};

export type BusinessSettings = {
  isStoreOpen: boolean;
  acceptingNewOrders: boolean;
  autoAcceptOrders: boolean;
  expressDelivery: boolean;
  pickupRadiusKm: number;
  openingTime: string;
  closingTime: string;
  weeklyOff: string;
  dailyOrderCap: number;
};

export type PartnerNotification = {
  id: string;
  title: string;
  body: string;
  time: string;
  read: boolean;
  kind: "order" | "payout" | "alert" | "promo";
};
